man/bntrackd.1
.\"
.\" Copyright (C) 2000  Ross Combs (ross@bnetd.org)
.\" 
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.TH BNPROXY 1 "12 September, 2000" "BNETD" "BNETD User's Manual"
.SH NAME
bnproxy \- bnetd tracking daemon
.SH SYNOPSIS
.B bntrackd
[\fIoptions\fP]
.BR
.SH DESCRIPTION
.B bntrackd
is a simple daemon which will listen for UDP announcement packets sent by
.BR bnetd (1)
servers.  The list of servers and their statistics is written to a text
file an optionally processed by an external program.  A
.BR perl (1)
script is
provided which processes this information to produce an HTML document
suitable for display on the web.
.LP
.BR bnetd (1)
servers must be configured to report to tracking servers.  The
.BR bnetd.conf (5)
file allows for multiple tracking servers to be listed separated by commas.
The master server is track.bnetd.org.  Please consider including that server
in your configuration.
.SH OPTIONS
.TP
.B \-c --command
Run the specified command after writing server information to \fIoutfile\fP.
If this option is not specified, the command "scripts/process.pl" is executed
from the directory that
.B bntrackd
was started.  If an empty string is provided for \fIcommand\fP, no command
will be executed.
.TP
.B \-e --expire
Set the period of time in seconds to keep server information without receiving
new updates.  If this option is not specified, 600 seconds is used.
.TP
.B \-d --debug
Place additional informational messages in the event log.
.TP
.B \-h --help --usage
Print usage information and exit.
.TP
.B \-l --logfile
Put the event log messages into the specified file.  If this option is not
specified, the file "logs/bntrackd.log" will be written from the directory
that
.B bntrackd
was started.  If an empty string is provided for \fIlogfile\fP, no event
messages will be written.
.TP
.TP
.B \-o --outfile
Put the server information into the specified file \fIoutfile\fP..
.TP
.B \-p --port
Listen for announcements on the specified UDP port number.  If this option
is not specified, the port number defaults to 6114.
.TP
.B \-P --pidfile
Record the process ID of the tracking daemon in the specified file.  If this
option is not specified, this information is not written to any file.
.TP
.B \-u --update
Set the period of time in seconds between updates of the \fIoutfile\fP.  If
this option is not specified, 150 seconds is used.
.TP
.B \-v --version
Causes
.B bntrackd
to print its version number and exit.
.SH NOTES
.TP
The current version is available at:
.LP
.RS
.I http://www.bnetd.org/
.RE
.SH "SEE ALSO"
.BR bnetd (1),
.BR perl (1)
.SH AUTHOR
Ross Combs (\fIross@bnetd.org\fP)


NEWS

		    PvPGN 1.7.x Major Changes And New Features
		  ==============================================

 o Westwood Online (WOL) support

    This is a historic moment in the life of the PvPGN project. For the first 
    time a different gaming network than Battle.net is supported with PvPGN. 
    Current supports includes almost complete support for Tiberian Sun (TS) 
    and basic support for Red Alert 2 (RA2) and Yuri's Revenge (YR).

 o WIN32 GUI for d2cs and d2dbs

    d2cs and d2bs now feature a new slick GUI if build under WIN32.

 o Instant startup (partial)

    PvPGN now can startup in a lot lesser time than before, this is due to 
    code changes which make it not load all accounts at startup. This 
    improvement should be mostly noticed on very high number of accounts 
    servers. Note that currently this happens only to SQL storage types 
    (the file based storage types still load all accounts at startup).

 o Optimized account loading/saving codes

    Account loading/saving codes have been rewritten allowing PvPGN to scale 
    a lot better with a high number of online users (ie lower CPU usage).

 o Better game result calculation

    Game result decision has been completly changed to use a democratic 
    ellection scheme (ie it bases the decision on the oppinions of the most 
    results). This should solve problems with game results beeing wrong under 
    some situations (like team games).

 o New WIN32 compiler/enviroment support

    Now you can compile PvPGN under WIN32 using complete GPL/free build tools 
    by using Dev-C++ project and additional tools (mingw, etc).

 o Optimized handling of D2 clients

    Replaced a O(n) search algorithm through the client connection list used 
    mostly by D2 handling functions with a O(1) algorithm.

 o Optimized timers codes

    Timers codes have been rewritten so the most used functions complete a 
    lot faster than before. This improves PvPGN CPU usage a lot because some 
    timers functions are used very often inside PvPGN.

 o Memory management changes and optimizations

    There has been added some memory management wrappers functions which 
    never fail and thus the calling codes were simplified a lot by removing 
    checks to memory allocation failures. If memory allocation fails PvPGN 
    tries to recover by programming an immediate shutdown and cleaning up 
    some safe memory.

 o Changed codes to use logged in username instead of account username

    This change makes PvPGN use the username introduced at login (with the 
    exact case as in login) instead of the account name. This solves some 
    bugs (like could not edit profile in D2 if logged in with different case) 
    and also gets us more close to Battle.Net.

 o Switched from string based tags to integer based tags

    Codes have been dramatically changed to use integers for storing of tags 
    (client tags, architecture tags, etc) and thus when comparing tags we use 
    plain integer comparation instead of string comparation. This should 
    reduce CPU usage a lot.

 o Optimized list handling codes

    The list API has been changed and the codes have been optimized so it 
    will remove elements from the list upon request and not later, thus 
    eliminating a O(n) search and thus making PvPGN a lot faster.

 o Unified and extended address translation system

    We have unified all address translation files and codes to use a single 
    address_translation.conf file for this job. Also the file syntax has 
    been completly changed to be more flexible and allow configuration of 
    almost any thinkable NAT setup.

 o IRC support reworked

    IRC client support has been bugfixed, optimized and made to work with a 
    lot more IRC clients.

 o Unified file based storage

    File based storage (plain and cdb) have been unified into a single storage 
    backend (called "file") and the mode (plain/cdb) is just a parameter.

 o Rewritten low level socket multiplexing codes (fdwatch2)

    The codes which deal with how PvPGN knows which sockets are available for 
    reading/writing have been rewritten with a new interface which does not 
    presume (like the old codes) that the underlying OS allocates socket 

src/bntrackd/bntrackd.cpp
/*
 * Copyright (C) 1999  Mark Baysinger (mbaysing@ucsd.edu)
 * Copyright (C) 2000  Ross Combs (rocombs@cs.nmsu.edu)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include "common/setup_before.h"
#define TRACKER_INTERNAL_ACCESS

#include <cstdio>
#include <cstring>
#include <ctime>
#include <cstdlib>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#include "compat/stdfileno.h"
#include "compat/psock.h"
#include "compat/inet_ntoa.h"
#include "compat/pgetpid.h"
#include "common/tracker.h"
#include "common/eventlog.h"
#include "common/list.h"
#include "common/xalloc.h"
#include "common/util.h"
#include "common/version.h"
#include "common/bn_type.h"
#include "common/setup_after.h"

using namespace pvpgn;

namespace {

	/******************************************************************************
	 * TYPES
	 *****************************************************************************/
	typedef struct
	{
		struct in_addr address;
		std::time_t         updated;
		t_trackpacket  info;
	} t_server;

	typedef struct
	{
		int            foreground;
		int            debug;
		int            XML_mode;
		unsigned int   expire;
		unsigned int   update;
		unsigned short port;
		char const *   outfile;
		char const *   pidfile;
		char const *   logfile;
		char const *   process;
	} t_prefs;


	/******************************************************************************
	 * STATIC FUNCTION PROTOTYPES
	 *****************************************************************************/
	int server_process(int sockfd);
	void usage(char const * progname);
	void getprefs(int argc, char * argv[]);
	void fixup_str(bn_byte * str, unsigned int size);


	/******************************************************************************
	 * GLOBAL VARIABLES
	 *****************************************************************************/
	t_prefs prefs;
	t_list *           serverlist_head;
}


extern int main(int argc, char * argv[])
{
	int sockfd;
	int result;

	if (argc < 1 || !argv || !argv[0])
	{
		std::fprintf(stderr, "bad arguments\n");
		return EXIT_FAILURE;
	}

	getprefs(argc, argv);

	if (!prefs.debug)
		eventlog_del_level("debug");
	if (prefs.logfile)
	{
		eventlog_set(stderr);
		if (eventlog_open(prefs.logfile) < 0)
		{
			eventlog(eventlog_level_fatal, __FUNCTION__, "could not use file \"%s\" for the eventlog (exiting)", prefs.logfile);
			return EXIT_FAILURE;
		}
	}

#ifdef DO_DAEMONIZE
	if (!prefs.foreground)
	{
		switch (fork())
		{
		case -1:
			eventlog(eventlog_level_error, __FUNCTION__, "could not fork (fork: %s)\n", std::strerror(errno));
			return EXIT_FAILURE;
		case 0: /* child */
			break;
		default: /* parent */
			return EXIT_SUCCESS;
		}

		close(STDINFD);
		close(STDOUTFD);
		close(STDERRFD);

# ifdef HAVE_SETPGID
		if (setpgid(0, 0) < 0)
		{
			eventlog(eventlog_level_error, __FUNCTION__, "could not create new process group (setpgid: %s)\n", std::strerror(errno));
			return EXIT_FAILURE;
		}
# else
#  ifdef HAVE_SETPGRP
#   ifdef SETPGRP_VOID
		if (setpgrp() < 0)
		{
			eventlog(eventlog_level_error, __FUNCTION__, "could not create new process group (setpgrp: %s)\n", std::strerror(errno));
			return EXIT_FAILURE;
		}
#   else
		if (setpgrp(0, 0) < 0)
		{
			eventlog(eventlog_level_error, __FUNCTION__, "could not create new process group (setpgrp: %s)\n", std::strerror(errno));
			return EXIT_FAILURE;
		}
#   endif
#  else
#   ifdef HAVE_SETSID
		if (setsid() < 0)
		{
			eventlog(eventlog_level_error, __FUNCTION__, "could not create new process group (setsid: %s)\n", std::strerror(errno));
			return EXIT_FAILURE;
		}
#   else
#    error "One of setpgid(), setpgrp(), or setsid() is required"
#   endif
#  endif
# endif
	}
#endif

	if (prefs.pidfile)
	{
#ifdef HAVE_GETPID
		std::FILE * fp;

		if (!(fp = std::fopen(prefs.pidfile, "w")))
		{
			eventlog(eventlog_level_error, __FUNCTION__, "unable to open pid file \"%s\" for writing (std::fopen: %s)", prefs.pidfile, std::strerror(errno));
			prefs.pidfile = NULL;
		}
		else
		{
			std::fprintf(fp, "%u", (unsigned int)getpid());
			if (std::fclose(fp) < 0)
				eventlog(eventlog_level_error, __FUNCTION__, "could not close pid file \"%s\" after writing (std::fclose: %s)", prefs.pidfile, std::strerror(errno));
		}
#else
		eventlog(eventlog_level_warn, __FUNCTION__, "no getpid() system call, do not use the -P or the --pidfile option");
		prefs.pidfile = NULL;
#endif
	}

#ifdef HAVE_GETPID
	eventlog(eventlog_level_info, __FUNCTION__, "bntrackd version "PVPGN_VERSION" process %u", (unsigned int)getpid());
#else
	eventlog(eventlog_level_info, __FUNCTION__, "bntrackd version "PVPGN_VERSION);
#endif

	if (psock_init() < 0)
	{
		eventlog(eventlog_level_error, __FUNCTION__, "could not initialize socket functions");
		return EXIT_FAILURE;
	}

	/* create the socket */
	if ((sockfd = psock_socket(PSOCK_PF_INET, PSOCK_SOCK_DGRAM, PSOCK_IPPROTO_UDP)) < 0)
	{
		eventlog(eventlog_level_error, __FUNCTION__, "could not create UDP listen socket (psock_socket: %s)\n", std::strerror(psock_errno()));
		return EXIT_FAILURE;
	}

	{
		struct sockaddr_in servaddr;

		/* bind the socket to correct port and interface */
		std::memset(&servaddr, 0, sizeof(servaddr));
		servaddr.sin_family = PSOCK_AF_INET;
		servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
		servaddr.sin_port = htons(prefs.port);
		if (psock_bind(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0)
		{
			eventlog(eventlog_level_error, __FUNCTION__, "could not bind to UDP port %hu (psock_bind: %s)\n", prefs.port, std::strerror(psock_errno()));
			return EXIT_FAILURE;
		}
	}

	if (!(serverlist_head = list_create()))
	{
		eventlog(eventlog_level_error, __FUNCTION__, "could not create server list");
		return EXIT_FAILURE;;
	}


	result = server_process(sockfd);


	if (serverlist_head != NULL){
		list_destroy(serverlist_head);
	}

	if (result < 0)
		return EXIT_FAILURE;

	return EXIT_SUCCESS;
}

namespace {

	int server_process(int sockfd)
	{
		t_elem *           curr;
		t_server *         server;
		struct sockaddr_in cliaddr;
		t_psock_fd_set     rfds;
		struct timeval     tv;
		std::time_t             last;
		std::FILE *             outfile;
		psock_t_socklen    len;
		t_trackpacket      packet;

		/* the main loop */
		last = std::time(NULL) - prefs.update;
		for (;;)
		{
			/* time to dump our list to disk and call the process command */
			/* (I'm making the assumption that this won't take very long.) */
			if (last + (signed)prefs.update < std::time(NULL))
			{
				last = std::time(NULL);

				if (!(outfile = std::fopen(prefs.outfile, "w")))
				{
					eventlog(eventlog_level_error, __FUNCTION__, "unable to open file \"%s\" for writing (std::fopen: %s)", prefs.outfile, std::strerror(errno));
					continue;
				}

				LIST_TRAVERSE(serverlist_head, curr)
				{
					server = (t_server*)elem_get_data(curr);

					if (server->updated + (signed)prefs.expire < last)
					{
						list_remove_elem(serverlist_head, &curr);
						xfree(server);
					}
					else
					{
						if (prefs.XML_mode == 1)
						{
							std::fprintf(outfile, "<server>\n\t<address>%s</address>\n", inet_ntoa(server->address));
							std::fprintf(outfile, "\t<port>%hu</port>\n", bn_short_nget(server->info.port));
							std::fprintf(outfile, "\t<location>%s</location>\n", server->info.server_location);
							std::fprintf(outfile, "\t<software>%s</software>\n", server->info.software);
							std::fprintf(outfile, "\t<version>%s</version>\n", server->info.version);
							std::fprintf(outfile, "\t<users>%lu</users>\n", bn_int_nget(server->info.users));
							std::fprintf(outfile, "\t<channels>%lu</channels>\n", bn_int_nget(server->info.channels));
							std::fprintf(outfile, "\t<games>%lu</games>\n", bn_int_nget(server->info.games));
							std::fprintf(outfile, "\t<description>%s</description>\n", server->info.server_desc);
							std::fprintf(outfile, "\t<platform>%s</platform>\n", server->info.platform);
							std::fprintf(outfile, "\t<url>%s</url>\n", server->info.server_url);
							std::fprintf(outfile, "\t<contact_name>%s</contact_name>\n", server->info.contact_name);
							std::fprintf(outfile, "\t<contact_email>%s</contact_email>\n", server->info.contact_email);
							std::fprintf(outfile, "\t<uptime>%lu</uptime>\n", bn_int_nget(server->info.uptime));
							std::fprintf(outfile, "\t<total_games>%lu</total_games>\n", bn_int_nget(server->info.total_games));
							std::fprintf(outfile, "\t<logins>%lu</logins>\n", bn_int_nget(server->info.total_logins));
							std::fprintf(outfile, "</server>\n");
						}
						else
						{
							std::fprintf(outfile, "%s\n##\n", inet_ntoa(server->address));
							std::fprintf(outfile, "%hu\n##\n", bn_short_nget(server->info.port));
							std::fprintf(outfile, "%s\n##\n", server->info.server_location);
							std::fprintf(outfile, "%s\n##\n", server->info.software);
							std::fprintf(outfile, "%s\n##\n", server->info.version);
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.users));
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.channels));
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.games));
							std::fprintf(outfile, "%s\n##\n", server->info.server_desc);
							std::fprintf(outfile, "%s\n##\n", server->info.platform);
							std::fprintf(outfile, "%s\n##\n", server->info.server_url);
							std::fprintf(outfile, "%s\n##\n", server->info.contact_name);
							std::fprintf(outfile, "%s\n##\n", server->info.contact_email);
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.uptime));
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.total_games));
							std::fprintf(outfile, "%lu\n##\n", bn_int_nget(server->info.total_logins));
							std::fprintf(outfile, "###\n");
						}
					}
				}
				if (std::fclose(outfile) < 0)
					eventlog(eventlog_level_error, __FUNCTION__, "could not close output file \"%s\" after writing (std::fclose: %s)", prefs.outfile, std::strerror(errno));

				if (prefs.process[0] != '\0')
					std::system(prefs.process);
			}

			/* select socket to operate on */
			PSOCK_FD_ZERO(&rfds);
			PSOCK_FD_SET(sockfd, &rfds);
			tv.tv_sec = BNTRACKD_GRANULARITY;
			tv.tv_usec = 0;
			switch (psock_select(sockfd + 1, &rfds, NULL, NULL, &tv))
			{
			case -1: /* error */
				if (
#ifdef PSOCK_EINTR
					errno != PSOCK_EINTR &&
#endif
					1)
					eventlog(eventlog_level_error, __FUNCTION__, "select failed (select: %s)", std::strerror(errno));
			case 0: /* timeout and no sockets ready */
				continue;
			}

			/* New tracking packet */
			if (PSOCK_FD_ISSET(sockfd, &rfds))
			{

				len = sizeof(cliaddr);
				if (psock_recvfrom(sockfd, &packet, sizeof(packet), 0, (struct sockaddr *)&cliaddr, &len) >= 0)
				{

					if (bn_short_nget(packet.packet_version) >= TRACK_VERSION)
					{
						bn_byte_set(&packet.software[sizeof(packet.software) - 1], '\0');
						fixup_str(packet.software, sizeof(packet.software));
						bn_byte_set(&packet.version[sizeof(packet.version) - 1], '\0');
						fixup_str(packet.version, sizeof(packet.version));
						bn_byte_set(&packet.platform[sizeof(packet.platform) - 1], '\0');
						fixup_str(packet.platform, sizeof(packet.platform));
						bn_byte_set(&packet.server_desc[sizeof(packet.server_desc) - 1], '\0');
						fixup_str(packet.server_desc, sizeof(packet.server_desc));
						bn_byte_set(&packet.server_location[sizeof(packet.server_location) - 1], '\0');
						fixup_str(packet.server_location, sizeof(packet.server_location));
						bn_byte_set(&packet.server_url[sizeof(packet.server_url) - 1], '\0');
						fixup_str(packet.server_url, sizeof(packet.server_url));
						bn_byte_set(&packet.contact_name[sizeof(packet.contact_name) - 1], '\0');
						fixup_str(packet.contact_name, sizeof(packet.contact_name));
						bn_byte_set(&packet.contact_email[sizeof(packet.contact_email) - 1], '\0');
						fixup_str(packet.contact_email, sizeof(packet.contact_email));

						/* Find this server's slot */
						LIST_TRAVERSE(serverlist_head, curr)
						{
							server = (t_server*)elem_get_data(curr);

							if (!std::memcmp(&server->address, &cliaddr.sin_addr, sizeof(struct in_addr)))
							{
								if (bn_int_nget(packet.flags)&TF_SHUTDOWN)
								{
									list_remove_elem(serverlist_head, &curr);
									xfree(server);
								}
								else
								{
									/* update in place */
									server->info = packet;
									server->updated = std::time(NULL);
								}
								break;
							}
						}

						/* Not found? Make a new slot */
						if (!(bn_int_nget(packet.flags)&TF_SHUTDOWN) && !curr)
						{
							server = (t_server*)xmalloc(sizeof(t_server));
							server->address = cliaddr.sin_addr;
							server->info = packet;
							server->updated = std::time(NULL);

							list_append_data(serverlist_head, server);
						}

						eventlog(eventlog_level_debug, __FUNCTION__,
							"Packet received from %s:"
							" packet_version=%u"
							" flags=0x%08lx"
							" port=%hu"
							" software=\"%s\""
							" version=\"%s\""
							" platform=\"%s\""
							" server_desc=\"%s\""
							" server_location=\"%s\""
							" server_url=\"%s\""
							" contact_name=\"%s\""
							" contact_email=\"%s\""
							" uptime=%lu"
							" total_games=%lu"
							" total_logins=%lu",
							inet_ntoa(cliaddr.sin_addr),
							bn_short_nget(packet.packet_version),
							bn_int_nget(packet.flags),
							bn_short_nget(packet.port),
							packet.software,
							packet.version,
							packet.platform,
							packet.server_desc,
							packet.server_location,
							packet.server_url,
							packet.contact_name,
							packet.contact_email,
							bn_int_nget(packet.uptime),
							bn_int_nget(packet.total_games),
							bn_int_nget(packet.total_logins));
					}

				}
			}

		}
	}


	void usage(char const * progname)
	{
		std::fprintf(stderr, "usage: %s [<options>]\n", progname);
		std::fprintf(stderr,
			"  -c COMMAND, --command=COMMAND  execute COMMAND update\n"
			"  -d, --debug                    turn on debug mode\n"
			"  -e SECS, --expire SECS         forget a list entry after SEC seconds\n"
#ifdef DO_DAEMONIZE
			"  -f, --foreground               don't daemonize\n"
#else
			"  -f, --foreground               don't daemonize (default)\n"
#endif
			"  -l FILE, --logfile=FILE        write event messages to FILE\n"
			"  -o FILE, --outfile=FILE        write server list to FILE\n");
		std::fprintf(stderr,
			"  -p PORT, --port=PORT           listen for announcments on UDP port PORT\n"
			"  -P FILE, --pidfile=FILE        write pid to FILE\n"
			"  -u SECS, --update SECS         write output file every SEC seconds\n"
			"  -x, --XML                      write output file in XML format\n"
			"  -h, --help, --usage            show this information and exit\n"
			"  -v, --version                  print version number and exit\n");
		std::exit(EXIT_FAILURE);
	}


	void getprefs(int argc, char * argv[])
	{
		int a;

		prefs.foreground = 0;
		prefs.debug = 0;
		prefs.expire = 0;
		prefs.update = 0;
		prefs.port = 0;
		prefs.XML_mode = 0;
		prefs.outfile = NULL;
		prefs.pidfile = NULL;
		prefs.process = NULL;
		prefs.logfile = NULL;

		for (a = 1; a < argc; a++)
		if (std::strncmp(argv[a], "--command=", 10) == 0)
		{
			if (prefs.process)
			{
				std::fprintf(stderr, "%s: processing command was already specified as \"%s\"\n", argv[0], prefs.process);
				usage(argv[0]);
			}
			prefs.process = &argv[a][10];
		}
		else if (std::strcmp(argv[a], "-c") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.process)
			{
				std::fprintf(stderr, "%s: processing command was already specified as \"%s\"\n", argv[0], prefs.process);
				usage(argv[0]);
			}
			a++;
			prefs.process = argv[a];
		}
		else if (std::strcmp(argv[a], "-d") == 0 || std::strcmp(argv[a], "--debug") == 0)
			prefs.debug = 1;
		else if (std::strncmp(argv[a], "--expire=", 9) == 0)
		{
			if (prefs.expire)
			{
				std::fprintf(stderr, "%s: expiration period was already specified as \"%u\"\n", argv[0], prefs.expire);
				usage(argv[0]);
			}
			if (str_to_uint(&argv[a][9], &prefs.expire) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], &argv[a][9]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-e") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.expire)
			{
				std::fprintf(stderr, "%s: expiration period was already specified as \"%u\"\n", argv[0], prefs.expire);
				usage(argv[0]);
			}
			a++;
			if (str_to_uint(argv[a], &prefs.expire) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], argv[a]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-f") == 0 || std::strcmp(argv[a], "--foreground") == 0)
			prefs.foreground = 1;
		else if (std::strcmp(argv[a], "-x") == 0 || std::strcmp(argv[a], "--XML") == 0)
			prefs.XML_mode = 1;
		else if (std::strncmp(argv[a], "--logfile=", 10) == 0)
		{
			if (prefs.logfile)
			{
				std::fprintf(stderr, "%s: eventlog file was already specified as \"%s\"\n", argv[0], prefs.logfile);
				usage(argv[0]);
			}
			prefs.logfile = &argv[a][10];
		}
		else if (std::strcmp(argv[a], "-l") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.logfile)
			{
				std::fprintf(stderr, "%s: eventlog file was already specified as \"%s\"\n", argv[0], prefs.logfile);
				usage(argv[0]);
			}
			a++;
			prefs.logfile = argv[a];
		}
		else if (std::strncmp(argv[a], "--outfile=", 10) == 0)
		{
			if (prefs.outfile)
			{
				std::fprintf(stderr, "%s: output file was already specified as \"%s\"\n", argv[0], prefs.outfile);
				usage(argv[0]);
			}
			prefs.outfile = &argv[a][10];
		}
		else if (std::strcmp(argv[a], "-o") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.outfile)
			{
				std::fprintf(stderr, "%s: output file was already specified as \"%s\"\n", argv[0], prefs.outfile);
				usage(argv[0]);
			}
			a++;
			prefs.outfile = argv[a];
		}
		else if (std::strncmp(argv[a], "--pidfile=", 10) == 0)
		{
			if (prefs.pidfile)
			{
				std::fprintf(stderr, "%s: pid file was already specified as \"%s\"\n", argv[0], prefs.pidfile);
				usage(argv[0]);
			}
			prefs.pidfile = &argv[a][10];
		}
		else if (std::strncmp(argv[a], "--port=", 7) == 0)
		{
			if (prefs.port)
			{
				std::fprintf(stderr, "%s: port number was already specified as \"%hu\"\n", argv[0], prefs.port);
				usage(argv[0]);
			}
			if (str_to_ushort(&argv[a][7], &prefs.port) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], &argv[a][7]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-p") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.port)
			{
				std::fprintf(stderr, "%s: port number was already specified as \"%hu\"\n", argv[0], prefs.port);
				usage(argv[0]);
			}
			a++;
			if (str_to_ushort(argv[a], &prefs.port) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], argv[a]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-P") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.pidfile)
			{
				std::fprintf(stderr, "%s: pid file was already specified as \"%s\"\n", argv[0], prefs.pidfile);
				usage(argv[0]);
			}
			a++;
			prefs.pidfile = argv[a];
		}
		else if (std::strncmp(argv[a], "--update=", 9) == 0)
		{
			if (prefs.update)
			{
				std::fprintf(stderr, "%s: update period was already specified as \"%u\"\n", argv[0], prefs.expire);
				usage(argv[0]);
			}
			if (str_to_uint(&argv[a][9], &prefs.update) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], &argv[a][9]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-u") == 0)
		{
			if (a + 1 >= argc)
			{
				std::fprintf(stderr, "%s: option \"%s\" requires an argument\n", argv[0], argv[a]);
				usage(argv[0]);
			}
			if (prefs.update)
			{
				std::fprintf(stderr, "%s: update period was already specified as \"%u\"\n", argv[0], prefs.expire);
				usage(argv[0]);
			}
			a++;
			if (str_to_uint(argv[a], &prefs.update) < 0)
			{
				std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], argv[a]);
				usage(argv[0]);
			}
		}
		else if (std::strcmp(argv[a], "-h") == 0 || std::strcmp(argv[a], "--help") == 0 || std::strcmp(argv[a], "--usage")
			== 0)
			usage(argv[0]);
		else if (std::strcmp(argv[a], "-v") == 0 || std::strcmp(argv[a], "--version") == 0)
		{
			std::printf("version "PVPGN_VERSION"\n");
			std::exit(0);
		}
		else if (std::strcmp(argv[a], "--command") == 0 || std::strcmp(argv[a], "--expire") == 0 ||
			std::strcmp(argv[a], "--logfile") == 0 || std::strcmp(argv[a], "--outfile") == 0 ||
			std::strcmp(argv[a], "--port") == 0 || std::strcmp(argv[a], "--pidfile") == 0 ||
			std::strcmp(argv[a], "--update") == 0)
		{
			std::fprintf(stderr, "%s: option \"%s\" requires and argument.\n", argv[0], argv[a]);
			usage(argv[0]);
		}
		else
		{
			std::fprintf(stderr, "%s: unknown option \"%s\"\n", argv[0], argv[a]);
			usage(argv[0]);
		}

		if (!prefs.process)
			prefs.process = BNTRACKD_PROCESS;
		if (prefs.expire == 0)
			prefs.expire = BNTRACKD_EXPIRE;
		if (!prefs.logfile)
			prefs.logfile = BNTRACKD_LOGFILE;
		if (!prefs.outfile)
			prefs.outfile = BNTRACKD_OUTFILE;
		if (prefs.port == 0)
			prefs.port = BNTRACKD_SERVER_PORT;
		if (!prefs.pidfile)
			prefs.pidfile = BNTRACKD_PIDFILE;
		if (prefs.expire == 0)
			prefs.update = BNTRACKD_UPDATE;

		if (prefs.logfile[0] == '\0')
			prefs.logfile = NULL;
		if (prefs.pidfile[0] == '\0')
			prefs.pidfile = NULL;
	}


	void fixup_str(bn_byte * str, unsigned int size)
	{
		bn_byte         prev;
		unsigned int i;

		bn_byte_set(&prev, '\0');

		for (i = 0; i < size, bn_byte_get(str[i]) != '\0'; bn_byte_set(&prev, bn_byte_get(str[i])), i++)
		if (bn_byte_get(prev) == '#' && bn_byte_get(str[i]) == '#')
			bn_byte_set(&str[i], '%');
	}

}
